pythonPackages.i3ipc: init at 1.4.0

vanzef opened the pull request:
closes #34828

###### Motivation for this change
[Request](https://github.com/NixOS/nixpkgs/issues/34828) from @cypher1.


###### Things done

Added package `pythonPackages.i3ipc`.

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

vanzef committed:
pythonPackages.i3ipc: init at 1.4.0
closes #34828

dotlambda commented:
Please rebase.

dotlambda commented:
@GrahamcOfBorg build python2.pkgs.i3ipc python3.pkgs.i3ipc

GrahamcOfBorg commented:
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.34835&attempt_id=bdc38fca-e3dc-4252-8da5-7fa5600876e2)

<details><summary>Partial log (click to expand)</summary><p>

```
reading manifest file 'i3ipc.egg-info/SOURCES.txt'
writing manifest file 'i3ipc.egg-info/SOURCES.txt'
running build_ext

----------------------------------------------------------------------
Ran 0 tests in 0.000s

OK
/nix/store/f1ncvwl638i6kygz2jkbv33ipxbhk2mm-python2.7-i3ipc-1.4.0
/nix/store/i3lrfjafbm5adj6qf6pwz8wanic3wr14-python3.6-i3ipc-1.4.0
```
</p></details>

GrahamcOfBorg commented:
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.34835&attempt_id=90c88145-6559-4463-89cb-478332cbd87e)

<details><summary>Partial log (click to expand)</summary><p>

```
reading manifest file 'i3ipc.egg-info/SOURCES.txt'
writing manifest file 'i3ipc.egg-info/SOURCES.txt'
running build_ext

----------------------------------------------------------------------
Ran 0 tests in 0.000s

OK
/nix/store/gfz6ijp6fr0fj4zbwj2psjfik1489kfg-python2.7-i3ipc-1.4.0
/nix/store/fxxl7y0v457v4xa3i5cag717a2a384pk-python3.6-i3ipc-1.4.0
```
</p></details>

FRidh commented:
```
Ran 0 tests in 0.000s
```
No tests are found. Either the test runner cannot find tests and needs to be patched, or there simply are no tests in which case the tests need to be disabled. Do include a comment explaining why the tests are disabled.

dotlambda commented:
> Ran 0 tests in 0.000s
Please set `doCheck = false` and add a comment why they are disabled. I think we would need to run an X server to run the tests.

vanzef commented:
@FRidh @dotlambda 
There are some tests in git repo, however there are none in PyPi package, so I've changed `src` to GitHub release, added `i3`, `xorgserver` (for xvfb) and `pytest` packages. Now tests are passing.

dotlambda commented on pkgs/development/python-modules/i3ipc/default.nix:
```diff
@@ -0,0 +1,27 @@
+
+  propagatedBuildInputs = [ enum-compat ];
+
+  checkPhase = ''python run-tests.py'';
```
`${python.interpreter}`

vanzef replied:
fixed

dotlambda commented:
@GrahamcOfBorg build python2.pkgs.i3ipc python3.pkgs.i3ipc

GrahamcOfBorg commented:
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.34835&attempt_id=37d2b017-6a90-4652-a8b7-8db96fda6e94)

<details><summary>Partial log (click to expand)</summary><p>

```
============================= test session starts ==============================
platform linux -- Python 3.6.4, pytest-3.4.0, py-1.5.2, pluggy-0.6.0
rootdir: /tmp/nix-build-python3.6-i3ipc-1.4.0.drv-0/i3ipc-python-1.4.0, inifile:
collected 2 items

test/test_window.py ..                                                   [100%]

=========================== 2 passed in 0.16 seconds ===========================
/nix/store/y9rdq2l9200yn80lig7051k46cnwnpip-python2.7-i3ipc-1.4.0
/nix/store/s9wixl7dlvkzbv5i732gynncvh7psvxx-python3.6-i3ipc-1.4.0
```
</p></details>

GrahamcOfBorg commented:
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.34835&attempt_id=34cb2a7b-af5a-4321-b15e-743fe3f12690)

<details><summary>Partial log (click to expand)</summary><p>

```
============================= test session starts ==============================
platform linux -- Python 3.6.4, pytest-3.4.0, py-1.5.2, pluggy-0.6.0
rootdir: /build/i3ipc-python-1.4.0, inifile:
collected 2 items

test/test_window.py ..                                                   [100%]

=========================== 2 passed in 0.36 seconds ===========================
/nix/store/n3ahr4c49z3casmw8sia76s2c6dgc819-python2.7-i3ipc-1.4.0
/nix/store/3zkmxghz0696ychwkdfxdfrkk06mmivx-python3.6-i3ipc-1.4.0
```
</p></details>

dotlambda commented on pkgs/development/python-modules/i3ipc/default.nix:
```diff
@@ -0,0 +1,27 @@
+  pname = "i3ipc";
+  version = "1.4.0";
+
+  src = fetchurl {
```
`fetchFromGitHub`

vanzef replied:
Is it really better? With `fetchurl` one can change only `version` and `sha256` when with `fetchFromGitHub` the `rev` also should be specified. Probably, there is a way to download by tags and not revision?

FRidh replied:
yes, `rev = tag`

dotlambda replied:
No, actually you'd have to change `url` and `sha256` with fetchurl. `rev` is definitely easier to change than `url`.

vanzef replied:
Ok, got it and fixed.

dotlambda commented on pkgs/development/python-modules/i3ipc/default.nix:
```diff
@@ -0,0 +1,27 @@
+    sha256 = "06x1zc9qizmcg4lxlh2w5amcq7igqlrs1kkvvi1czb5wy689ygsi";
+  };
+
+  nativeBuildInputs = [ xorgserver pytest i3 ];
```
This is probably better suited for `checkInputs`.

vanzef replied:
fixed

dotlambda commented:
@GrahamcOfBorg build python2.pkgs.i3ipc python3.pkgs.i3ipc

GrahamcOfBorg commented:
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.34835&attempt_id=8927b4ce-91cf-4057-a75a-e1d6b078595a)

<details><summary>Partial log (click to expand)</summary><p>

```
============================= test session starts ==============================
platform linux -- Python 3.6.4, pytest-3.4.0, py-1.5.2, pluggy-0.6.0
rootdir: /build/source, inifile:
collected 2 items

test/test_window.py ..                                                   [100%]

=========================== 2 passed in 0.35 seconds ===========================
/nix/store/fkabi7jf1janbwnq4miqz6pb0bn4x51c-python2.7-i3ipc-1.4.0
/nix/store/4wmix95jfdkxz7xv09p62psjaqgh7k7l-python3.6-i3ipc-1.4.0
```
</p></details>

GrahamcOfBorg commented:
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.34835&attempt_id=71ed64b2-5e38-4696-aece-85dfaef9830c)

<details><summary>Partial log (click to expand)</summary><p>

```
i3ipc/i3ipc.py:10: in <module>
    from enum import Enum
E   ImportError: No module named enum
!!!!!!!!!!!!!!!!!!! Interrupted: 1 errors during collection !!!!!!!!!!!!!!!!!!!!
=========================== 1 error in 0.31 seconds ============================
test/test_window.py ..                                                   [100%]

=========================== 2 passed in 0.75 seconds ===========================
/nix/store/w1r1q0cj4l8hhxnkkn56jdpfwsxil9ck-python2.7-i3ipc-1.4.0
/nix/store/rqqqs8x14d40mcr2ighrgmccjnf7p965-python3.6-i3ipc-1.4.0
```
</p></details>

dotlambda reviewed (approved):
Thanks a lot!

Cypher1 commented:
Thanks very much.
Is this process required for all python packages or can we automate it for most?

FRidh commented:
This is indeed a process that is required for all packages. 

Aside from https://github.com/NixOS/nixpkgs/blob/master/maintainers/scripts/update-python-libraries there is no automation for Python in Nixpkgs. There have been attempts, but because of various reasons there is no good solution for in Nixpkgs. For other projects `pypi2nix` may be a good solution though.
